### ISSUE TITLE
Setting up Slack integration not possible for multi-environment development

### ISSUE THREAD

**Issue by toddpadwick** [verbb/formie#791]
**Description**
I have tried to set up a Slack integration on my local development. However, my local development URL that is auto-generated by formie for the redirect URL is not a suitable URL allowed by Slack and does not allow the connection. I also need the URL to be the production URL. So the only way that i can could set it up to work is to allow admin changes in production and connect in prod. However, this has now resulted n my yaml files being out of sync with development. 
How do I get this to work locally before committing all changed to production?

This is what happens when I try to authorize locally:

![Screenshot 2022-03-08 at 10 25 35](https://user-images.githubusercontent.com/36191462/157207527-f1a839ac-80d3-4ff3-aef1-e1925b06d7a4.png)

![Screenshot 2022-03-08 at 10 26 29](https://user-images.githubusercontent.com/36191462/157207720-adeb2529-856c-486c-a6a3-5fd1df5a93d1.png)

**Comment by engram-design** [comment]
I can confirm this works with a real development URL, like `https://formie.test`, producing the URI `https://formie.test/actions/formie/integrations/callback`. You'll also want to ensure you enable `'usePathInfo' => true` in your general config, as some providers don't like having query strings in the URL. This URI isn't configurable.

You'd typically add all your URLs in the Slack app (development, staging, production) so that they all work. Authorizing locally during development will suffice for testing and setting up field mapping. When deploying to production, you can disconnect/reconnect to the integration - but it's not required. The OAuth handshake will detect when a token is out of date and request a new one.

Are you able to change your development URL to include a domain name? It seems to be `functionandform-api` which would be invalid.

**Comment by toddpadwick** [comment]
Hi @engram-design  I have updated the url to include a proper url with SSL and a dot instead of the hyphen - `https://functionandform.api`
I have added usePath info to config.
However I still get the following error when trying to connect:
![Screenshot 2022-03-19 at 15 29 24](https://user-images.githubusercontent.com/36191462/159127435-762ebbf3-d381-4f29-93e0-0a47ae92cfaa.png)
 `https://functionandform.api/`

**Comment by engram-design** [comment]
Is `.api` a valid TLD? I don't seem to have an issue connecting, and I'm not sure why it's flagging `Invalid client_id parameter`. You get that error when hitting "Connect" in the Formie integration, correct?

Here's my settings:

![Screen Shot 2022-03-20 at 08 56 49](https://user-images.githubusercontent.com/1221575/159139798-ae658bab-6a50-403d-a830-3e6e137e8ada.png)
![screencapture-api-slack-apps-A036M4T3D8S-2022-03-20-08_56_56](https://user-images.githubusercontent.com/1221575/159139808-afae5887-ff24-4669-9c0a-8e3e2f4e612c.png)
![screencapture-api-slack-apps-A036M4T3D8S-oauth-2022-03-20-08_58_21](https://user-images.githubusercontent.com/1221575/159139826-d3f63e8f-a98b-41b5-8007-f90ff9efbd66.png)

**Comment by toddpadwick** [comment]
Hi @engram-design  I can confirm the settings match, except we have two redirect URLs. one for prod which has already connected, and one for locale. And yes, it happens when clicking the connect button when in local dev.

**Comment by toddpadwick** [comment]
Hi @engram-design  I have got the app connected by simply re-doing the process from scratch, starting with local dev. 
However, despite local dev successfully returning slack channels, in production it doesn't. The loading wheel flickers and then nothing is returned. Can you confirm slack channels return successfully for you in multiple environments even if they are connected already in another?

![Screenshot 2022-03-23 at 10 48 02](https://user-images.githubusercontent.com/36191462/159671356-113209bd-d637-45a0-a51f-7e8303a0976c.png)

**Comment by engram-design** [comment]
Anything in your `storage/logs/formie.log` file? I have checked that on a different environment it does work. On production, can you try disconnecting and reconnecting from the Formie > Settings > Misc > Slack screen, just in case there's an issue with the production redirect.

However, it should be clearly showing an error if one has occurred.

Are you able to open up the Network tab in web developer tools (for Chrome), and click the refresh button? Look for the new XHR request that pops up in the Network tab, and click on it. It should look similar to the below

![image](https://user-images.githubusercontent.com/1221575/159821084-707bbed3-88b0-4990-8e6c-ca939b1a97f2.png)

**Comment by toddpadwick** [comment]
Sorry for delay on this - busy setting up a number of sites at the moment, but please keep this open. I'll look at the logs ASAP as requested

**Comment by toddpadwick** [comment]
Hi @engram-design 
I've discovered I might be getting a related issue on another site we're setting up salesforce on. 
Just like with slack, we managed to connect salesforce locally. But on production, when refreshing integration, we get this error:
```
Error: Request failed with status code 500
API error: “Attempt to read property "accessToken" on null” /srv/app/thewriter-cms/htdocs/vendor/verbb/formie/src/integrations/crm/Salesforce.php:432
```

Initially, i thought i must have forgotten to add the env var for the consumer secret and id but I triple checked and can confirm the keys have been added just as they have locally. I even checked in settings on prod to see if they were pulling through fine and they are:
<img width="1104" alt="Screenshot 2022-03-30 at 14 50 34" src="https://user-images.githubusercontent.com/36191462/160838319-71efc9bf-b9e3-4e49-b88a-4de2f394d4fc.png">


If you'd prefer me to create a separate ticket for this, let me know.


With regards to slack on the other site, there are no logs created – we're hosting with the [pro stack on Fortrabbit](https://help.fortrabbit.com/craft-3-start) – which is load balanced and therefore has no locally stored logs. The only logs I get are the global PHP and apache logs which have to be tailed - this is all I found relating to formie:

```
apache_access  91.126.213.35 - - [30/Mar/2022:12:55:31 +0000] "POST /index.php/admin/actions/formie/integrations/form-settings HTTP/1.1" 500 3517 "https://cms.functionandform.co.uk/admin/formie/forms/edit/5945" "Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/99.0.4844.83 Safari/537.36"
```

**Comment by engram-design** [comment]
What version of Formie are you on? I note you didn't include this, as this error was taken care of a few versions back in [1.5.11](https://github.com/verbb/formie/blob/craft-3/CHANGELOG.md#1511---2022-03-12).

As for the logs - ah I see, that's a bit of a pain! Were you able to see a response within the Network tab when refreshing the integration?

**Comment by toddpadwick** [comment]
Hey @engram-design Just updated the site with the slack integration to 1.5.12. Still having the same issue where the channels don't load. 
see below screenshot from network tab for formie/settings:
![Screenshot 2022-03-31 at 14 13 01](https://user-images.githubusercontent.com/36191462/161052110-69d326f9-3c1f-4aec-8cb7-ba435bc5302f.png)

Seems to be missing data in the response when compared with yours?

**Comment by engram-design** [comment]
Well at least it's not failing. This might be a silly question, but do you have any public channels in your Slack group? We will only show public channels according to https://github.com/verbb/formie/blob/618491d7c2fad482572d2c9c7f986c64f0a07c64/src/integrations/miscellaneous/Slack.php#L166-L172

**Comment by toddpadwick** [comment]
@engram-design  No... see list of channels that come up locally:

![Screenshot 2022-03-31 at 14 20 16](https://user-images.githubusercontent.com/36191462/161054097-b76a07f7-330b-4cbe-a32f-c85d031a2d0c.png)

but its not a silly question and does touch on another issue I've been trying to figure out:
The client has given us a temporary slack account for the purpose of connecting their slack. However, they need to remove our access asap due to sensitive information in their slack conversations that are not for us to see.

But by the looks of things, the slack connection is through a specific user and once our access is revoked, so will the connection. Considering this has to be done on local dev first, then on production, how do we go about this – Theres no way for me to get the client to connect their own account, on my local machine? by the sounds of it, it slack integrations may need to be done on each environment specifically, they may need to be removed from yaml, admin level settings, and into non-admin settings for this reason, allowing integrations to be made into accounts out of the control of the developer? What do you reckon?

**Comment by engram-design** [comment]
There must be something wrong with how your production hosting is calling services, but seems extremely odd. I may have to spin up a Fortrabbit instance just to be sure. I've not used them to-date, but I feel like they do one or two things out of the ordinary.

As for the connection, I'd get your client to setup the API details, to generate the client keys. Add them locally, and you won't be able to test the connection. Deploy these to the server, and have your client visit Formie > Settings > Misc > Slack and connect via OAuth. This will prompt them to login to their Slack account/organisation once to authentication OAuth. Once that's done, subsequent new OAuth tokens will be received. FYI, I'd also recommend storing API keys/etc as .env variable for easier swapping-out, and keeping sensitive info safe.

But yes, It's a bit of an odd situation where we can't have it both ways - being in project config and synced across environments, and have per-environment settings/configuration/connection/etc.

**Comment by toddpadwick** [comment]
Hi @engram-design  won't that result in the yaml files being ahead of local in production?  I noticed that the action of connecting is also stored in yaml (presumably the OAuth tokens), so if its not connected locally, but connected on prod they'll end up being overwritten next deployment right?

The keys are already env vars and they were created by the client – I initially tried to use my own slack locally, and there's on prod but that didn't solve the issue above.

**Comment by toddpadwick** [comment]
Hi @engram-design  regarding the failed load of Slack channels, I spoke with Fortrabbit and they came back with the following:

> First guess into the blue: Somehow related to ephemeral storage with our Pro Stack. It won't work if some configuration is written to a file from the App. You can store state in the database or in Memcache, but not on the local file system.
> More on ephemeral storage:
> ﻿https://help.fortrabbit.com/app-pro#toc-ephemeral-storage
> Again, that is just a guess without understanding full context. I have leveled this case up.

**Comment by toddpadwick** [comment]
> What version of Formie are you on? I note you didn't include this, as this error was taken care of a few versions back in [1.5.11](https://github.com/verbb/formie/blob/craft-3/CHANGELOG.md#1511---2022-03-12).
> 
> As for the logs - ah I see, that's a bit of a pain! Were you able to see a response within the Network tab when refreshing the integration?

With regards to Salesforce I have updated to latest formie - 1.5.12, and can confirm, I'm still getting an error – slightly different one this time though:
``` 
Error: Request failed with status code 500
API error: “verbb\formie\services\Tokens::refreshToken(): Argument #1 ($token) must be of type verbb\formie\models\Token, null given, called in /srv/app/thewriter-cms/htdocs/vendor/verbb/formie/src/integrations/crm/Salesforce.php on line 459” /srv/app/thewriter-cms/htdocs/vendor/verbb/formie/src/services/Tokens.php:140
```
(please let me know if you'd like me to move this salesforce issue over to another ticket)

**Comment by engram-design** [comment]
The connection result isn't store in YAML. If you have a look in your `formie_integrations` database table, there's a `cache` column that stores the connection state, and all the settings used by the form (in your case, channels, etc). This is per-environment.

OAuth tokens are purely db-specific (stored in `formie_tokens`).

As for the OAuth tokens, a reference to their `tokenId` is stored in YAML, but I can now see that will pose an issue. Each time you push it to production, it'll contain an invalid `tokenId` referencing your local tokens. 

I can confirm if I get my tokens out of sync on production I'll get your exact issues, so that sounds like what's going on.

Unfortunately, that leaves us in a tricky situation that poses breaking changes. I need to remove `tokenId` from the YAML files, but as soon as I do, it'll wipe out any `tokenId`'s set on production, breaking many people's integrations. I'll give this some further thought on a less breaking way to handle this.

But for now, I would recommend to push all your changes up to production, connect to all your integrations on production, and ensure the tokens are generated, and working correctly. Then download the database from production, so your local site has the same tokens as production (running `./craft project-config/rebuild`).

As I said, I'll be working on a fix for this in the meantime.

**Comment by engram-design** [comment]
For the moment, I've just made it so that `tokenId` values stored in YAML just aren't applied. That should hopefully fix this issue. Just be sure to manually re-connect to your integrations on production, or manually restore the `tokenId` column values in `formie_integrations` on production. Any further changes locally to the YAML files will be ignored. As I've mentioned, I can't remove them from YAML files, but they'll at least be ignored.

To get this fix early, change your `verbb/formie` requirement in `composer.json` to:

```
"require": {
  "verbb/formie": "dev-craft-3 as 1.5.12",
  "...": "..."
}
```

Then run `composer update`. Sorry for the hassle!

**Comment by toddpadwick** [comment]
Thanks @engram-design . I've updated formie as instructed, but when I try and go to the _formie/settings/miscellaneous_ I get this error:
![Screenshot 2022-04-01 at 15 01 16](https://user-images.githubusercontent.com/36191462/161269205-8cac0231-c944-4a04-b6c3-6daac04debb5.png)

**Comment by engram-design** [comment]
Sorry about that, should be fixed now if you run `composer update` again.

**Comment by toddpadwick** [comment]
Hi @engram-design  I have got that Slack integration connected now. and it partially works, however there is a new issue here: https://github.com/verbb/formie/issues/822

**Comment by engram-design** [comment]
Fixed in [1.5.13](https://github.com/verbb/formie/releases/tag/1.5.13) (will continue following up in other thread)
